docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,361 @@
+
+## Changelog
+
+- 2022-09-10: First draft (@thanethomson)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+This ADR proposes to effectively offload some data storage responsibilities and
+functionality to **single external companion service** in order to:
+
+1. Improve core consensus stability.
+2. Eventually reduce the amount of functionality for which the core team is
+   responsible, thereby improving maintainability.
+3. Still cater for certain use cases that require this data and functionality.
+
+The way the system is currently built is such that a Tendermint node is mostly
+self-contained. While this philosophy initially allowed a certain degree of ease
+of node operation (i.e. simple UX), it has also lent itself to feature sprawl,
+with Tendermint being asked to take care of increasingly more than just
+consensus. Under the spotlight in this ADR are:
+
+1. **Event indexing**, which is required in order to facilitate arbitrary block
+   and transaction querying, as well as subscription for arbitrary events. We
+   have already seen, for instance, how the current (somewhat unreliable) event
+   subscription implementation can cause back-pressure into consensus and affect
+   IBC relayer stability (see [\#6729] and [\#7156]).
+2. **Block execution result storage**, which is required to facilitate a number
+   of RPC APIs, but also results in storage of large quantities of data that is
+   not critical to consensus. This data is, however, critical for certain use
+   cases outside of consensus.
+
+Another intersecting issue is that there may be a variety of use cases that
+require different data models. A good example of this is the mere existence of
+the [PostgreSQL indexer] as compared to the default key/value indexer.
+
+Continuing from and expanding on the ideas outlined in [RFC 006][rfc-006], the
+suggested approach in this ADR is to provide a mechanism to publish certain data
+from Tendermint, in real-time and with certain reliability guarantees, to a
+single companion service outside of the node that can use this data in whatever
+way it chooses (filter and republish it, store it, manipulate or enrich it,
+etc.).
+
+Specifically, this mechanism would initially publish:
+
+1. The results of block execution (e.g. data from `FinalizeBlockResponse`). This
+   data is not accessible from the P2P layer, and currently provides valuable
+   information for Tendermint users (whether events should be handled at all by
+   Tendermint is a different problem).
+2. All block data (i.e. `Block` data structures that have been committed by the
+   consensus engine).
+
+## Alternative Approaches
+
+One clear alternative to this would be the approach outlined in [ADR
+075][adr-075]. This approach still unfortunately leaves Tendermint responsible
+for maintaining a query interface and event indexing functionality, increasing
+the long-term maintenance burden of, and the possibility of feature sprawl in,
+that subsystem.
```

Comment:
We could do that, but it doesn't solve all the problems this ADR aims to solve.

1. The solution in ADR-075 only keeps track of a sliding window of events (and does so in memory), and so there's still the possibility that subscribers who need guaranteed event delivery (e.g. relayers) may end up losing events in the case of intermittent network failures.
2. We would need to establish a way to impose a single-consumer limit in order to be able to automatically prune stored events and block results. The assumption in this ADR is that, because we have a single reliable companion, if we've sent them the data we won't need to re-send it.
3. We would still be offering the data via JSON as opposed to Protobuf. This may have performance and bandwidth implications with this "firehose" design.

If we just modified ADR-075 to address these issues, we would probably converge on the same approach outlined in this ADR except we'd be sending the data via JSON over HTTP instead of Protobuf over HTTP/2.

Reply:
I've updated the ADR to reflect these answers.